gef-ui/server/controllers/bank-review-date/post-bank-review-date.ts

```diff
@@ -0,0 +1,161 @@
+import * as api from '../../services/api';
+import { validateAndParseBankReviewDate } from './validation';
+import { asLoggedInUserSession, LoggedInUserSession } from '../../utils/express-session';
+import { getCoverStartDateOrStartOfToday } from '../../utils/get-cover-start-date-or-start-of-today';
+import { BankReviewDateViewModel } from '../../types/view-models/bank-review-date-view-model';
+import { Facility } from '../../types/facility';
+
+type BankReviewDateParams = { dealId: string; facilityId: string };
+type BankReviewDatePostBody = { 'bank-review-date-day': string; 'bank-review-date-month': string; 'bank-review-date-year': string };
+
+type PostBankReviewDateRequest = CustomExpressRequest<{
+  reqBody: BankReviewDatePostBody;
+  params: BankReviewDateParams;
+  query: { saveAndReturn: string; status: string | undefined };
+}>;
+
+type PostRequestUris = {
+  nextPage: string;
+  previousPage: string;
+  saveAndReturn: string;
+};
+
+type PostBankReviewDateParams = {
+  req: PostBankReviewDateRequest;
+  res: Response;
+  uris: PostRequestUris;
+};
+
+/**
+ * Update the bank review date if it has changed
+ * @param existingFacility the current facility from the api
+ * @param bankReviewDate the new bank review date
+ * @param session the user session
+ */
+const updateBankReviewDateIfChanged = async (existingFacility: Facility, bankReviewDate: Date, { userToken, user }: LoggedInUserSession): Promise<void> => {
+  const bankReviewDateNeedsUpdating =
+    typeof existingFacility.bankReviewDate !== 'string' || !isSameDay(parseISO(existingFacility.bankReviewDate), bankReviewDate);
+
+  if (!bankReviewDateNeedsUpdating) {
+    return;
+  }
+
+  await api.updateFacility({
+    facilityId: existingFacility._id,
+    payload: {
+      bankReviewDate,
+    },
+    userToken,
+  });
+
+  const applicationUpdate = {
+    editorId: user._id,
+  };
+  await api.updateApplication({ dealId: existingFacility.dealId, application: applicationUpdate, userToken });
+};
+
+/**
+ * Handle the bank review date request
+ */
+const postBankReviewDate = async ({ req, res, uris }: PostBankReviewDateParams) => {
+  try {
+    const {
+      params: { dealId, facilityId },
+      body: { 'bank-review-date-year': bankReviewDateYear, 'bank-review-date-month': bankReviewDateMonth, 'bank-review-date-day': bankReviewDateDay },
+      query: { saveAndReturn, status },
+    } = req;
+
+    const { userToken } = asLoggedInUserSession(req.session);
+
+    const bankReviewDateIsBlank = !bankReviewDateYear && !bankReviewDateMonth && !bankReviewDateDay;
+
+    // If the user clicks save and return with no values filled in, we do not update the database
+    if (isTrueSet(saveAndReturn) && bankReviewDateIsBlank) {
+      return res.redirect(uris.saveAndReturn);
+    }
+
+    const { details: facility } = await api.getFacility({ facilityId, userToken });
+
+    const bankReviewDateErrorsAndDate = validateAndParseBankReviewDate(
```

Comment:
Could/should we destructure the results here, thinking maybe:

```ts
const { validationErrors, submittedDate } = validateAndParseBankReviewDate()
```

Would simplify some instances below

Reply:
`validateAndParseBankReviewDate` returns a `ErrorsOrDate` object - we have to check if the `errors` object exists in the return value before we can use it

Reply:
Interesting approach - not a blocker since it's an established pattern, I think it could be simpler 😊